{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7cff75b1",
   "metadata": {},
   "source": [
    "Name: Lydia Holley\n",
    "\n",
    "Email: lholley4@uncc.edu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "958fc755",
   "metadata": {},
   "source": [
    "Your classes should, at minimum:\n",
    "- have a __repr__ and __str__that provide a meaningful representation as a string\n",
    "- check that the bases or amino acids in the string are valid\n",
    "- work as the argument for a SequenceRecord\n",
    "\n",
    "DNA\n",
    "- a translate method that will convert the DNA sequence and return a ProteinSequence object\n",
    "- one other method of your choice (what you did previously is fine)\n",
    "\n",
    "Protein\n",
    "- a method of your choice. In this case, if the method you would implement is too complex to reasonably implement or would use resources you don't have access to, it is okay to leave it as what is called a stub method (has only one line, \"pass\") and explain in comments what this method would do and it's purpose\n",
    " \n",
    "Here is a dictionary you can copy into your code to help facilitate DNA translation:\n",
    "aa_dict = {'M':['ATG'], 'F':['TTT', 'TTC'], 'L':['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'C':['TGT', 'TGC'], 'Y':['TAC', 'TAT'], 'W':['TGG'], 'P':['CCT', 'CCC', 'CCA', 'CCG'], 'H':['CAT', 'CAC'], 'Q':['CAA', 'CAG'], 'R':['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'I':['ATT', 'ATC', 'ATA'], 'T':['ACT', 'ACC', 'ACA', 'ACG'], 'N':['AAT', 'AAC'], 'K':['AAA', 'AAG'], 'S':['AGT', 'AGC', 'TCT', 'TCC', 'TCA', 'TCG'], 'V':['GTT', 'GTC', 'GTA', 'GTG'], 'A':['GCT', 'GCC', 'GCA', 'GCG'], 'D':['GAT', 'GAC'], 'E':['GAA', 'GAG'], 'G':['GGT', 'GGC', 'GGA', 'GGG'], '*':['TAA','TAG','TGA']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "bbc3276b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sequence class\n",
    "from functools import total_ordering\n",
    "\n",
    "@total_ordering\n",
    "class Sequence:\n",
    "     # initialize with valid characters\n",
    "    def __init__(self,seq):\n",
    "        valid_char = \"MFLCYWPHQRITNKSVADEG*\"\n",
    "        for i in range(len(seq)):\n",
    "            if seq[i] in valid_char:\n",
    "                continue\n",
    "            else:\n",
    "                raise ValueError(f'Error at position {i+1}')\n",
    "        self.seq = seq\n",
    "       \n",
    "    def __str__(self):\n",
    "        return self.seq\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'Sequence: {self.seq}'\n",
    "    \n",
    "    def __eq__(self,other): #this will only allow objects of the same class to be equal\n",
    "        if isinstance(other,DNAseq):\n",
    "            if isinstance(self,DNAseq) and self.seq == other.seq:\n",
    "                return True\n",
    "            else:\n",
    "                return False            \n",
    "        elif isinstance(other,ProteinSeq):\n",
    "            if isinstance(self,ProteinSeq) and self.seq == other.seq:\n",
    "                return True\n",
    "            else:\n",
    "                return False            \n",
    "        elif isinstance(other,Sequence):\n",
    "            if isinstance(self,DNAseq):\n",
    "                return False\n",
    "            elif isinstance(self,ProteinSeq):\n",
    "                return False\n",
    "            elif isinstance(self,Sequence) and self.seq == other.seq:\n",
    "                return True\n",
    "            else:\n",
    "                return False            \n",
    "        else:\n",
    "            return False\n",
    "        \n",
    "    def __add__(self,other): #this will only allow objects of the same class to be added to each other\n",
    "        if isinstance(other,DNAseq):\n",
    "            if isinstance(self,DNAseq):\n",
    "                return DNAseq(self.seq + other.seq)\n",
    "            else:\n",
    "                print(\"cannot add sequences of different classes\")\n",
    "                return False            \n",
    "        elif isinstance(other,ProteinSeq):\n",
    "            if isinstance(self,ProteinSeq):\n",
    "                return ProteinSeq(self.seq + other.seq)\n",
    "            else:\n",
    "                print(\"cannot add sequences of different classes\")\n",
    "                return False           \n",
    "        elif isinstance(other,Sequence):\n",
    "            if isinstance(self,Sequence):\n",
    "                return Sequence(self.seq + other.seq)\n",
    "            else:\n",
    "                print(\"cannot add sequences of different classes\")\n",
    "                return False           \n",
    "        else:\n",
    "            return False\n",
    "        \n",
    "    def __lt__(self,other):\n",
    "        if type(self.seq) == type(other.seq):\n",
    "            return len(self.seq) < len(other.seq)\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "582fd867",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DNA sequence class\n",
    "\n",
    "@total_ordering\n",
    "class DNAseq(Sequence):\n",
    "    \n",
    "    def __init__(self,seq):\n",
    "        valid_char = \"ATCG\"\n",
    "        for i in range(len(seq)):\n",
    "            if seq[i] in valid_char:\n",
    "                continue\n",
    "            else:\n",
    "                raise ValueError(f'Invalid base at position {i+1}')\n",
    "        super().__init__(seq)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return f'DNA Sequence: {self.seq}'\n",
    "    \n",
    "    def translate_to_protein(self): #translates DNA to Amino Acids. Utilized last semester notes as an aid\n",
    "        aa_dict = {'M':['ATG'], 'F':['TTT', 'TTC'], 'L':['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'C':['TGT', 'TGC'], 'Y':['TAC', 'TAT'], 'W':['TGG'], 'P':['CCT', 'CCC', 'CCA', 'CCG'], 'H':['CAT', 'CAC'], 'Q':['CAA', 'CAG'], 'R':['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'I':['ATT', 'ATC', 'ATA'], 'T':['ACT', 'ACC', 'ACA', 'ACG'], 'N':['AAT', 'AAC'], 'K':['AAA', 'AAG'], 'S':['AGT', 'AGC', 'TCT', 'TCC', 'TCA', 'TCG'], 'V':['GTT', 'GTC', 'GTA', 'GTG'], 'A':['GCT', 'GCC', 'GCA', 'GCG'], 'D':['GAT', 'GAC'], 'E':['GAA', 'GAG'], 'G':['GGT', 'GGC', 'GGA', 'GGG'], '*':['TAA','TAG','TGA']}\n",
    "        protein_seq = \"\"\n",
    "        for i in range(0,len(self.seq),3):\n",
    "            codon = self.seq[i:i+3]\n",
    "            for aa,codons in aa_dict.items():\n",
    "                if codon in codons:\n",
    "                    protein_seq += aa\n",
    "                    \n",
    "        return ProteinSeq(protein_seq)\n",
    "    \n",
    "    def __GC_content__(self): #produces the proportion of the sequence that is G and C base pairs\n",
    "        GC_count = 0\n",
    "        for base in self.seq:\n",
    "            if base == \"G\" or base == \"C\":\n",
    "                GC_count += 1\n",
    "            else:\n",
    "                continue\n",
    "        proportionGC = GC_count/len(self.seq)\n",
    "        return proportionGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "1fa2a048",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FML\n",
      "LYDIA\n"
     ]
    }
   ],
   "source": [
    "#confirms the translation works\n",
    "seq2 = DNAseq(\"TTCATGCTT\")\n",
    "print(seq2.translate_to_protein()) #FML\n",
    "seqL = DNAseq(\"TTATACGACATCGCC\")\n",
    "print(seqL.translate_to_protein()) #LYDIA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "318a1247",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Protein Sequence Class\n",
    "\n",
    "@total_ordering\n",
    "class ProteinSeq(Sequence):\n",
    "    \n",
    "    def __init__(self,seq):        \n",
    "        valid_char = \"MFLCYWPHQRITNKSVADEG*\"\n",
    "        for i in range(len(seq)):\n",
    "            if seq[i] in valid_char:\n",
    "                continue\n",
    "            else:\n",
    "                raise ValueError(f'Invalid Amino Acid at position {i+1}')\n",
    "        super().__init__(seq)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return f'Protein Sequence: {self.seq}'\n",
    "    \n",
    "    def __highest_AA_content__(self): #returns the amino acid found most in the sequence\n",
    "        aa_counter = {}\n",
    "        aa_highest = []\n",
    "        for aa in self.seq:\n",
    "            if aa in aa_counter.keys():\n",
    "                aa_counter[aa] += 1\n",
    "            else:\n",
    "                aa_counter[aa] = 1\n",
    "        highest_value = max(aa_counter.values())\n",
    "        for aa,value in aa_counter.items():\n",
    "            if value == highest_value:\n",
    "                aa_highest.append(aa)\n",
    "            \n",
    "        return aa_highest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "6d0ae2d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATCGGCATGCATGCCTCGTAGCGTA\n",
      "ATCGGCATGCATGCCTCGTAGCGTA\n",
      "GTTCATGATGTACATGAATGATAC\n",
      "ATCGGCATGCATGCCTCGTAGCGTA\n",
      "NTEAGCYWINKSTHQRVD\n",
      "\n",
      "\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "True\n",
      "\n",
      "\n",
      "<class 'str'>\n",
      "<class '__main__.Sequence'>\n",
      "<class '__main__.DNAseq'>\n",
      "<class '__main__.DNAseq'>\n",
      "<class '__main__.ProteinSeq'>\n",
      "\n",
      "\n",
      "GTTCATGATGTACATGAATGATACATCGGCATGCATGCCTCGTAGCGTA\n",
      "GTTCATGATGTACATGAATGATAC\n",
      "cannot add sequences of different classes\n",
      "False\n",
      "cannot add sequences of different classes\n",
      "False\n",
      "NTEAGCYWINKSTHQRVDNWTEIWNKSTWAGCYWITEIWNNKSTHWQRYWINTEINNKKWVD\n",
      "\n",
      "\n",
      "0.3333333333333333\n",
      "0.56\n",
      "0.56\n",
      "\n",
      "\n",
      "['N', 'T']\n",
      "['W']\n"
     ]
    }
   ],
   "source": [
    "#assign variables to classes\n",
    "seq = \"ATCGGCATGCATGCCTCGTAGCGTA\"\n",
    "seq1 = Sequence(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq2 = DNAseq(\"GTTCATGATGTACATGAATGATAC\")\n",
    "seq3 = DNAseq(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq4 = ProteinSeq(\"NTEAGCYWINKSTHQRVD\")\n",
    "seq6 = DNAseq(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq7 = ProteinSeq(\"NWTEIWNKSTWAGCYWITEIWNNKSTHWQRYWINTEINNKKWVD\")\n",
    "\n",
    "#print variables to ensure proper output\n",
    "print(seq)  #ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seq1) #ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seq2) #GTTCATGATGTACATGAATGATAC\n",
    "print(seq3) #ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seq4) #NTEAGCYWINKSTHQRVD\n",
    "\n",
    "print(\"\\n\") #ensure that the __eq__ method works properly\n",
    "print(seq == seq1)  #False\n",
    "print(seq1 == seq2) #False\n",
    "print(seq1 == seq3) #False\n",
    "print(seq2 == seq3) #False\n",
    "print(seq1 == seq4) #False\n",
    "print(seq2 == seq4) #False\n",
    "print(seq3 == seq6) #True\n",
    "\n",
    "print(\"\\n\") #ensure that each variable is stored as the correct class object\n",
    "print(type(seq))  #str\n",
    "print(type(seq1)) #Sequence\n",
    "print(type(seq2)) #DNAseq\n",
    "print(type(seq3)) #DNAseq\n",
    "print(type(seq4)) #ProteinSeq\n",
    "\n",
    "print(\"\\n\") #checks that the __add__ method works properly\n",
    "print(seq2.__add__(seq3)) #GTTCATGATGTACATGAATGATACATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seq2)  #GTTCATGATGTACATGAATGATAC\n",
    "print(seq2.__add__(seq4)) #cannot add sequences of different classes. False\n",
    "print(seq2 + seq4) #cannot add sequences of different classes. False\n",
    "print(seq4 + seq7) #NTEAGCYWINKSTHQRVDNWTEIWNKSTWAGCYWITEIWNNKSTHWQRYWINTEINNKKWVD\n",
    "\n",
    "print(\"\\n\") #ensures that the __GC_content__ method works properly\n",
    "print(seq2.__GC_content__()) #0.3333333333333333\n",
    "print(seq3.__GC_content__()) #0.56\n",
    "print(seq6.__GC_content__()) #0.56\n",
    "\n",
    "print(\"\\n\") #ensures that highest amino acid count method works\n",
    "print(seq4.__highest_AA_content__()) #['N', 'T']\n",
    "print(seq7.__highest_AA_content__()) #['W']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "01f05214",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Error at position 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/2125354417.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#test whether sequence validation worked\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mseq70\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"???\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#this should return an error with a message about error position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/3895683606.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, seq)\u001b[0m\n\u001b[1;32m     11\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Error at position {i+1}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Error at position 1"
     ]
    }
   ],
   "source": [
    "#test whether sequence validation worked\n",
    "seq70 = Sequence(\"???\") #this should return an error with a message about error position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "ccf4eb32",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Invalid Amino Acid at position 2",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/4242758089.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#test whether protein seq validation worked\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mseq71\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mProteinSeq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"A??\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#this should return an error with a message about error position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/3637256514.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, seq)\u001b[0m\n\u001b[1;32m     10\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Invalid Amino Acid at position {i+1}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Invalid Amino Acid at position 2"
     ]
    }
   ],
   "source": [
    "#test whether protein seq validation worked\n",
    "seq71 = ProteinSeq(\"A??\") #this should return an error with a message about error position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "87b10c23",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Invalid base at position 3",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/3588953374.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#test whether DNA seq validation worked\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mseq72\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDNAseq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"AA?\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#this should return an error with a message about error position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/n6/gbphvvd12sq8410bcnt0m5pc0000gn/T/ipykernel_35442/3960838769.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, seq)\u001b[0m\n\u001b[1;32m     10\u001b[0m                 \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Invalid base at position {i+1}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Invalid base at position 3"
     ]
    }
   ],
   "source": [
    "#test whether DNA seq validation worked\n",
    "seq72 = DNAseq(\"AA?\") #this should return an error with a message about error position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "5c3b497b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SequenceRecord class just to check that the classes work as input\n",
    "from functools import total_ordering\n",
    "\n",
    "@total_ordering\n",
    "class SequenceRecord:\n",
    "\n",
    "    def __init__(self,title,*args):\n",
    "        self.title = title\n",
    "        for seq in args:\n",
    "            if isinstance(seq,Sequence):\n",
    "                self.sequence = seq\n",
    "            elif isinstance(seq,str):\n",
    "                new_seq = Sequence(seq)\n",
    "                self.sequence = new_seq\n",
    "  \n",
    "    def __str__(self):\n",
    "        return str(self.title)+ \"\\n\" + str(self.sequence)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'{self.title}: {self.sequence}'\n",
    "    \n",
    "    def __eq__(self,other):\n",
    "        if self.sequence == other.sequence:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    def __lt__(self,other):\n",
    "        return len(self.sequence) < len(other.sequence)\n",
    "    \n",
    "    def __gt__(self,other):\n",
    "        return len(self.sequence) > len(other.sequence)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "48c68550",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence 1: ATCGGCATGCATGCCTCGTAGCGTA\n",
      "Sequence 1: ATCGGCATGCATGCCTCGTAGCGTA\n",
      "<class '__main__.SequenceRecord'>\n",
      "Sequence 2: GTTCATGATGTACATGAATGATAC\n",
      "Sequence 2: GTTCATGATGTACATGAATGATAC\n",
      "<class '__main__.SequenceRecord'>\n"
     ]
    }
   ],
   "source": [
    "# Use this cell to test that the classes work as SequenceRecord class input\n",
    "\n",
    "seq = \"ATCGGCATGCATGCCTCGTAGCGTA\"\n",
    "seq1 = Sequence(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq2 = DNAseq(\"GTTCATGATGTACATGAATGATAC\")\n",
    "seq3 = DNAseq(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq4 = ProteinSeq(\"NTEAGCYWINKSTHQRVD\")\n",
    "seq6 = DNAseq(\"ATCGGCATGCATGCCTCGTAGCGTA\")\n",
    "seq7 = ProteinSeq(\"NWTEIWNKSTWAGCYWITEIWNNKSTHWQRYWINTEINNKKWVD\")\n",
    "\n",
    "seqRec_1 = SequenceRecord(\"Sequence 1\", seq1)\n",
    "print(repr(seqRec_1)) #Sequence 1: ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seqRec_1.__repr__()) #Sequence 1: ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(type(seqRec_1)) #<class '__main__.SequenceRecord'>\n",
    "\n",
    "seqRec_2 = SequenceRecord(\"Sequence 2\", seq2)\n",
    "print(repr(seqRec_2)) #Sequence 1: ATCGGCATGCATGCCTCGTAGCGTA\n",
    "print(seqRec_2.__repr__()) #Sequence 2: GTTCATGATGTACATGAATGATAC\n",
    "print(type(seqRec_2)) #<class '__main__.SequenceRecord'>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
